test/framework/src/main/java/org/elasticsearch/test/disruption/NetworkDisruption.java

```diff
@@ -0,0 +1,468 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.test.disruption;
+
+import com.carrotsearch.randomizedtesting.generators.RandomPicks;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.test.InternalTestCluster;
+import org.elasticsearch.test.transport.MockTransportService;
+import org.elasticsearch.transport.ConnectTransportException;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Random;
+import java.util.Set;
+
+import static org.junit.Assert.assertFalse;
+
+/**
+ * Network disruptions are modeled using two components:
+ * 1) the {@link DisruptedLinks} represents the links in the network that are to be disrupted
+ * 2) the {@link NetworkLinkDisruptionType} represents the failure mode that is to be applied to the links
+ */
+public class NetworkDisruption implements ServiceDisruptionScheme {
+
+    private final ESLogger logger = Loggers.getLogger(NetworkDisruption.class);
+
+    private final DisruptedLinks disruptedLinks;
+    private final NetworkLinkDisruptionType networkLinkDisruptionType;
+
+    protected volatile InternalTestCluster cluster;
+    protected volatile boolean activeDisruption = false;
+
+    public NetworkDisruption(DisruptedLinks disruptedLinks, NetworkLinkDisruptionType networkLinkDisruptionType) {
+        this.disruptedLinks = disruptedLinks;
+        this.networkLinkDisruptionType = networkLinkDisruptionType;
+    }
+
+    @Override
+    public void applyToCluster(InternalTestCluster cluster) {
+        this.cluster = cluster;
+    }
+
+    @Override
+    public void removeFromCluster(InternalTestCluster cluster) {
+        stopDisrupting();
+    }
+
+    @Override
+    public void removeAndEnsureHealthy(InternalTestCluster cluster) {
+        removeFromCluster(cluster);
+        ensureNodeCount(cluster);
+    }
+
+    protected void ensureNodeCount(InternalTestCluster cluster) {
+        assertFalse("cluster failed to form after disruption was healed", cluster.client().admin().cluster().prepareHealth()
+            .setWaitForNodes("" + cluster.size())
+            .setWaitForRelocatingShards(0)
+            .get().isTimedOut());
+    }
+
+    @Override
+    public synchronized void applyToNode(String node, InternalTestCluster cluster) {
+
+    }
+
+    @Override
+    public synchronized void removeFromNode(String node1, InternalTestCluster cluster) {
+        if (disruptedLinks.nodes().contains(node1)) {
```

Comment:
Can you do something about the duplicated deeply nested code here and in `startDisrupting`?